### PR TITLE
Refactor sphinx.environment; Add EnvironmentCollectors

### PR DESCRIPTION
**Pull request by tk0miya**
This will bring BuildEnvironment class simple.

- [x] API doc
- [x] convert domains to collectors
- [x] convert managers to collectors

### CODE REVIEW

**Comment by tk0miya** [comment]
>convert domains to collectors

I feel the domains are a bit complicated to convert collectors simply.
At this moment, it brings inconvenience and too much segmentalized modules.
Even if we will modify them, we have to refactor them more simply at first.

So I'd like to keep as it is (at least now).

**Comment by tk0miya** [comment]
I just finished all my work. This ready to merge.
@shimizukawa could you review this please?

**Comment by tk0miya** [comment]
Updated to resolve conflicts

File: CHANGES [hunk 1]
```diff
@@ -62,6 +62,14 @@ Deprecated
 * ``Sphinx.status_iterator()` and ``Sphinx.old_status_iterator()`` is now
   deprecated.  Please use ``sphinx.util:status_iterator()`` intead.
 * ``BuildEnvironment.set_warnfunc()`` is now deprecated
+* Following methods of ``BuildEnvironment`` is not deprecated.
```
Comment (shimizukawa) [inline CHANGES]: > is not deprecated.

*not* ?

**Review by shimizukawa — APPROVED** [review]
LGTM with nits ;)

File: CHANGES (hunk 1 above)
Comment (tk0miya) [reply CHANGES]: oh, it's typo.

**Comment by tk0miya** [comment]
Merged. Thank you for reviewing!

**Pull request merged**